vendor: docker/libnetwork b3507428be5b458cb0e2b4086b13531fb0706e46

thaJeztah opened the pull request:
also somewhat related to https://github.com/moby/moby/issues/28589, which discusses the "implicit" binding of IPv6

depends on:

- [x] https://github.com/moby/moby/pull/42021 Update rootlesskit to v0.13.1 to fix handling of IPv6 addresses
    - [x] https://github.com/rootless-containers/rootlesskit/pull/213 Refactor ParsePortSpec to handle IPv6 addresses, and improve validation
- [x] https://github.com/moby/moby/pull/42050 CI: update tests to be more resilient against CLI output format and for libnetwork changes
- [x] ~https://github.com/moby/moby/pull/42102~ https://github.com/moby/moby/pull/42179 update rootlesskit to v0.14.0
    - [x] https://github.com/rootless-containers/rootlesskit/pull/232 Port API: support specifying IP version explicitly ("tcp4", "tcp6")

full diff: https://github.com/docker/libnetwork/compare/fa125a3512ee0f6187721c88582bf8c4378bd4d7...b3507428be5b458cb0e2b4086b13531fb0706e46

- fixed IPv6 iptables rules for enabled firewalld (https://github.com/moby/libnetwork/pull/2609)
    - fixes https://github.com/moby/moby/issues/41861 "Docker uses 'iptables' instead of 'ip6tables' for IPv6 NAT rule, crashes" 
- Fix regression in https://github.com/moby/libnetwork/pull/2608
    - introduced in "Fix IPv6 Port Forwarding for the Bridge Driver" (https://github.com/moby/libnetwork/pull/2604)
    - fixes https://github.com/moby/libnetwork/issues/2607 "IPv4 and IPv6 addresses are not bound by default anymore"
    - fixes https://github.com/moby/moby/issues/41858 "IPv6 is no longer proxied by default anymore"
- Use hostIP to decide on Portmapper version (https://github.com/moby/libnetwork/pull/2616)
    - fixes docker-proxy not being stopped correctly

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
vendor: docker/libnetwork b3507428be5b458cb0e2b4086b13531fb0706e46

full diff: https://github.com/docker/libnetwork/compare/fa125a3512ee0f6187721c88582bf8c4378bd4d7...b3507428be5b458cb0e2b4086b13531fb0706e46

- fixed IPv6 iptables rules for enabled firewalld (libnetwork#2609)
    - fixes "Docker uses 'iptables' instead of 'ip6tables' for IPv6 NAT rule, crashes"
- Fix regression in docker-proxy
    - introduced in "Fix IPv6 Port Forwarding for the Bridge Driver" (libnetwork#2604)
    - fixes/addresses: "IPv4 and IPv6 addresses are not bound by default anymore" (libnetwork#2607)
    - fixes/addresses "IPv6 is no longer proxied by default anymore" (moby#41858)
- Use hostIP to decide on Portmapper version
    - fixes docker-proxy not being stopped correctly

Port mapping of containers now contain separatet mappings for IPv4 and IPv6 addresses, when
listening on "any" IP address. Various tests had to be updated to take multiple mappings into
account.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@arkodg PTAL; can you check if I have the correct "fixes" in place or if I missed any / linked the wrong ones?

thaJeztah commented:
Lots of failures; look related / legit failures;

```
=== FAIL: arm64.integration.container TestNetworkLoopbackNat (0.28s)
    nat_test.go:66: assertion failed: error is not nil: Error response from daemon: driver failed programming external connectivity on endpoint server-TestNetworkLoopbackNat (5a4eaaf3ff84ad959dc875a4829767a89dd3da6bc472e3f6feba32241b54d982): Bind for :::8080 failed: port is already allocated
```

```
=== RUN   TestNetworkLocalhostTCPNat
--- FAIL: TestNetworkLocalhostTCPNat (0.15s)
    nat_test.go:48: assertion failed: error is not nil: Error response from daemon: driver failed programming external connectivity on endpoint server-TestNetworkLocalhostTCPNat (09844ee44b347259889cac7df6f0a9ac78af25af4a690166d4e404eed2d01845): Error starting userland proxy: error while calling PortManager.AddPort(): listen tcp: address :::8081: too many colons in address
```

```
=== RUN   TestDockerSuite/TestPortExposeHostBinding
    --- FAIL: TestDockerSuite/TestPortExposeHostBinding (0.54s)
        docker_cli_port_test.go:314: assertion failed: expression is false: err == nil: out: 0.0.0.0:49153
            :::49153
```

```
=== RUN   TestDockerSuite/TestPortHostBinding
   --- FAIL: TestDockerSuite/TestPortHostBinding (0.53s)
       docker_cli_port_test.go:293: assertion failed: error is not nil: different size lists 0.0.0.0:9876
           :::9876
           , 2, 1
```

```
=== RUN   TestDockerSuite/TestPsNotShowPortsOfStoppedContainer
    --- FAIL: TestDockerSuite/TestPsNotShowPortsOfStoppedContainer (0.50s)
        docker_cli_ps_test.go:603: assertion failed: :::5000->5000/tcp (fields[len(fields)-2] string) != 0.0.0.0:5000->5000/tcp (expected string): Expected: 0.0.0.0:5000->5000/tcp, got: :::5000->5000/tcp
```

```
=== RUN   TestDockerSuite/TestRunAllocatePortInReservedRange
    --- FAIL: TestDockerSuite/TestRunAllocatePortInReservedRange (0.80s)
        docker_cli_run_test.go:2071: invalid port, got: 49155
            , error: strconv.ParseInt: parsing "49155\n": invalid syntax
```

```
=== RUN   TestDockerSuite/TestRunAllowPortRangeThroughExpose
    --- FAIL: TestDockerSuite/TestRunAllowPortRangeThroughExpose (0.71s)
        docker_cli_run_test.go:2282: Port is not mapped for the port 3000/tcp
```

thaJeztah commented:
@bboehmke PTAL

bboehmke commented:
1. failure:
```
=== FAIL: arm64.integration.container TestNetworkLoopbackNat (0.28s)
    nat_test.go:66: assertion failed: error is not nil: Error response from daemon: driver failed programming external connectivity on endpoint server-TestNetworkLoopbackNat (5a4eaaf3ff84ad959dc875a4829767a89dd3da6bc472e3f6feba32241b54d982): Bind for :::8080 failed: port is already allocated
```
is caused because the `docker-proxy` process for IPv6 is not removed after the container is stopped.

I think this is caused because the `releasePort` selects the wrong portMapper.

On creation the `bnd.IP` is used: https://github.com/moby/libnetwork/blob/master/drivers/bridge/port_mapping.go#L142
On release the `bnd.HostIP` is used: https://github.com/moby/libnetwork/blob/master/drivers/bridge/port_mapping.go#L212

For the IPv6 userland proxy the `bnd.HostIP` is `::` and the `bnd.IP` is a IPv4.

bboehmke commented:
failure 3
```
=== RUN   TestDockerSuite/TestPortExposeHostBinding
    --- FAIL: TestDockerSuite/TestPortExposeHostBinding (0.54s)
        docker_cli_port_test.go:314: assertion failed: expression is false: err == nil: out: 0.0.0.0:49153
            :::49153
```
failure 4
```
=== RUN   TestDockerSuite/TestPortHostBinding
   --- FAIL: TestDockerSuite/TestPortHostBinding (0.53s)
       docker_cli_port_test.go:293: assertion failed: error is not nil: different size lists 0.0.0.0:9876
           :::9876
           , 2, 1
```
failure 5
```
=== RUN   TestDockerSuite/TestPsNotShowPortsOfStoppedContainer
    --- FAIL: TestDockerSuite/TestPsNotShowPortsOfStoppedContainer (0.50s)
        docker_cli_ps_test.go:603: assertion failed: :::5000->5000/tcp (fields[len(fields)-2] string) != 0.0.0.0:5000->5000/tcp (expected string): Expected: 0.0.0.0:5000->5000/tcp, got: :::5000->5000/tcp
```

I think they are caused because the container now also shows the IPv6 publish (previously only IPv4).
Here an example:
```
root@docker-test:~# docker ps
CONTAINER ID   IMAGE     COMMAND                  CREATED         STATUS         PORTS                               NAMES
2eea9954bb0a   nginx     "/docker-entrypoint.…"   9 seconds ago   Up 8 seconds   0.0.0.0:80->80/tcp, :::80->80/tcp   nginx_test
```

bboehmke commented:
failure 7:
```
=== RUN   TestDockerSuite/TestRunAllowPortRangeThroughExpose
    --- FAIL: TestDockerSuite/TestRunAllowPortRangeThroughExpose (0.71s)
        docker_cli_run_test.go:2282: Port is not mapped for the port 3000/tcp
```
similar to the ones before but this time it expects that the binding is only correct if there is exactly one binding:
https://github.com/moby/moby/blob/master/integration-cli/docker_cli_run_test.go#L2281

The inspect output would look like this:
```json
            "Ports": {
                "80/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "80"
                    },
                    {
                        "HostIp": "::",
                        "HostPort": "80"
                    }
                ]
            },

```

bboehmke commented:
Sadly for the second one I have no idea what is going on.
Maybe it is also related to the first one but I can not test this locally because the integration test is not running on my system.

Sadly I don't think I will have the time to fix this on my one in the next days. 
But at least I hope this information helps to fix the issues/tests.

arkodg commented:
thanks for triaging the issues @bboehmke !
@thaJeztah can you please include https://github.com/moby/libnetwork/pull/2616 in this PR ?

The port bindings have now been split up into 2 - for ipv4 and ipv6
1. This helps set up port mapping using `iptables` for ipv4 bindings and `ip6tables` for ipv6 bindings in a cleaner way
2. earlier a binding using something like `-p 3000:3000` set up 2 bindings if using `userland-proxy` and 1 if using
`iptables` .
3. To resolve the tests failures, we can either look for 2 bindings - for `0.0.0.0` and `::` or disable userland proxy & ipv6

thaJeztah commented:
> The port bindings have now been split up into 2 - for ipv4 and ipv6

Looks like we should at least mention that in the changelog. Also /cc @simonferquel @djs55 as I think Docker Desktop makes use of that information to setup networking

thaJeztah commented:
> @thaJeztah can you please include moby/libnetwork#2616 in this PR ?

updated the commit to include that PR. Will have a look later at fixing the tests

thaJeztah commented:
Looks like this failure is an issue in rootless kit;

```
=== RUN   TestNetworkNat
--- FAIL: TestNetworkNat (0.22s)
    nat_test.go:30: assertion failed: error is not nil: Error response from daemon: driver failed programming external connectivity on endpoint server-TestNetworkNat (65cc966cfc2fd79fcfaf0063f3afa7174e8854ce9dfce04e7f9cd4cd0b1e0268): Error starting userland proxy: error while calling PortManager.AddPort(): listen tcp: address :::8080: too many colons in address
```

Probably somewhere in this function, which doesn't look to be handling IPv6 addresses; https://github.com/rootless-containers/rootlesskit/blob/27c9611a3931149dd91fabd2d559140c351bf36f/pkg/port/portutil/portutil.go#L13-L15

thaJeztah commented:
I thought I'd have a look at the rootless kit code (thought; replace with the go-connections code used in Docker and Libnetwork), but I think this change; https://github.com/rootless-containers/rootlesskit/pull/207 is problematic; the "port spec" string format is already ambiguous in docker (`[ip-address]:[port or port-range]:[port or port-range][/protocol]`), but could still be handled (knowing that `port` or `port-range` would always be either "empty", "integer", or "integer-integer"). Having _two_ IP addresses in the format makes things a lot more complicated if IPv6 is added to the equation (using `[]` could somewhat help for `::`, but `[::]`, or `2001:db8::1` I think would still be problematic.
Can (for now) probably work around it by making sure we use the IPv4 for that test. (test also looks to not put `[]` around IPv6 addresses, which probably should, as its `<ip>:<port>`)

thaJeztah commented:
moved to 'draft', pending https://github.com/moby/moby/pull/42050

thaJeztah commented:
Getting close; ppc64le failure is unrelated (need to create a ticket for that), but rootless failure looks related, and fails on 3 tests;

```
=== FAIL: amd64.integration.container TestNetworkNat (0.23s)
    nat_test.go:30: assertion failed: error is not nil: Error response from daemon:
    driver failed programming external connectivity on endpoint server-TestNetworkNat (cc45c612eb285bb1b52152305cf3d155ee7c547e2d1967575b4bc271d3be81f3):
    Error starting userland proxy:
    error while calling PortManager.AddPort(): listen tcp 0.0.0.0:8080: bind: address already in use

=== FAIL: amd64.integration.container TestNetworkLocalhostTCPNat (0.13s)
    nat_test.go:48: assertion failed: error is not nil: Error response from daemon:
    driver failed programming external connectivity on endpoint server-TestNetworkLocalhostTCPNat (90323fcd8d614dfd8cdc69df6727dd9d9a54911c7a12db548adef4e31bbb42cb):
    Error starting userland proxy:
    error while calling PortManager.AddPort(): listen tcp 0.0.0.0:8081: bind: address already in use

=== FAIL: amd64.integration.container TestNetworkLoopbackNat (0.13s)
    nat_test.go:66: assertion failed: error is not nil: Error response from daemon:
    driver failed programming external connectivity on endpoint server-TestNetworkLoopbackNat (81d16c3d777b5723524da166ecb20e04f7305a1415499fd579035f7ca201916c):
    Error starting userland proxy:
    error while calling PortManager.AddPort(): listen tcp 0.0.0.0:8080: bind: address already in use
```

Wondering if it's an issue in rootless kit (failing to distinguish `[::]:8080` from `0.0.0.0:8080`, or libnetwork not releasing the port (but in that case, I'd expect it to fail in non-rootless as well)

thaJeztah commented:
@AkihiroSuda do you have an idea what causes the failure on rootless? (port already in use)? Could it be it's somewhere conflating ipv4 and ipv6 (and trying to bind twice on ipv4?).

I tried checking if the ports were in use at the start of the tests (they weren't), so it happens when starting the container that's starred in the test)

AkihiroSuda commented:
```console
$ docker --context=rootless run --rm -p 8080:80 nginx:alpine
docker: Error response from daemon: driver failed programming external connectivity on endpoint vigorous_archimedes (8ca94868817b1f0398ae1c96aa4caa3eedc60753b826797ce164fa33a159a966): Error starting userland proxy: error while calling PortManager.AddPort(): listen tcp 0.0.0.0:8080: bind: address already in use.
```

```console
$ sudo execsnoop.bt 
Attaching 2 probes...
TIME(ms)   PID   ARGS
22491      43131 docker --context=rootless run --rm -p 8080:80 nginx:alpine
22542      43141 /sbin/iptables --wait -t nat -C DOCKER -p tcp -d 0/0 --dport 8080 -j DNAT --to-destination 172.17.0.2:80
22545      43142 /sbin/iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 8080 -j DNAT --to-destination 172.17.0.2:80
22547      43143 /sbin/iptables --wait -t filter -C DOCKER ! -i docker0 -o docker0 -p tcp -d 172.17.0.2 --dport
22548      43144 /sbin/iptables --wait -t filter -A DOCKER ! -i docker0 -o docker0 -p tcp -d 172.17.0.2 --dport
22550      43145 /sbin/iptables --wait -t nat -C POSTROUTING -p tcp -s 172.17.0.2 -d 172.17.0.2 --dport 80 -j MASQUERADE
22551      43146 /sbin/iptables --wait -t nat -A POSTROUTING -p tcp -s 172.17.0.2 -d 172.17.0.2 --dport 80 -j MASQUERADE
22553      43147 /usr/local/bin/rootlesskit-docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 8080 -container-ip 172.17.0.2 -container-port 80
22555      43152 docker-proxy -container-ip 172.17.0.2 -container-port 80 -host-ip 127.0.0.1 -host-port 8080 -proto tcp
22556      43160 /usr/local/bin/rootlesskit-docker-proxy -proto tcp -host-ip :: -host-port 8080 -container-ip 172.17.0.2 -container-port 80
22564      43167 /sbin/iptables --wait -t nat -C DOCKER -p tcp -d 0/0 --dport 8080 -j DNAT --to-destination 172.17.0.2:80
22568      43168 /sbin/iptables --wait -t nat -D DOCKER -p tcp -d 0/0 --dport 8080 -j DNAT --to-destination 172.17.0.2:80
22571      43169 /sbin/iptables --wait -t filter -C DOCKER ! -i docker0 -o docker0 -p tcp -d 172.17.0.2 --dport
22573      43170 /sbin/iptables --wait -t filter -D DOCKER ! -i docker0 -o docker0 -p tcp -d 172.17.0.2 --dport
22574      43171 /sbin/iptables --wait -t nat -C POSTROUTING -p tcp -s 172.17.0.2 -d 172.17.0.2 --dport 80 -j MASQUERADE
22576      43172 /sbin/iptables --wait -t nat -D POSTROUTING -p tcp -s 172.17.0.2 -d 172.17.0.2 --dport 80 -j MASQUERADE
```

`rootlesskit-docker-proxy` is unexpectedly called twice:
```
/usr/local/bin/rootlesskit-docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 8080 -container-ip 172.17.0.2 -container-port 80
/usr/local/bin/rootlesskit-docker-proxy -proto tcp -host-ip :: -host-port 8080 -container-ip 172.17.0.2 -container-port 80
```

Seems a bug on libnetwork side.

arkodg commented:
@AkihiroSuda the behavior in libnetwork has recently changed to spawn a separate `docker-proxy` process for each port binding type - ipv4 & ipv6, this might look non optimal when in userland-proxy mode, but enables creating  correct port forwarding rules for `ipv4` and `ipv6` bindings using `iptables` & `ip6tables` respectively

AkihiroSuda commented:
Opened https://github.com/rootless-containers/rootlesskit/issues/231 for discussion

AkihiroSuda commented:
Please try this version of RootlessKit: https://github.com/rootless-containers/rootlesskit/pull/232

thaJeztah commented:
included https://github.com/moby/moby/pull/42102 to test

thaJeztah commented:
Rebased; this should be ready for review: @AkihiroSuda @tonistiigi @tiborvass @cpuguy83 ptal

thaJeztah committed:
fix assertPortList normalizing being too strict

The normalizing was updated with the output of the "docker port" command
in mind, but we're normalizing the "expected" output, which is passed
without the "->" in front of the mapping, causing some tests to fail;

    === RUN   TestDockerSuite/TestPortHostBinding
        --- FAIL: TestDockerSuite/TestPortHostBinding (1.21s)
            docker_cli_port_test.go:324: assertion failed: error is not nil: |:::9876!=[::]:9876|
    === RUN   TestDockerSuite/TestPortList
        --- FAIL: TestDockerSuite/TestPortList (0.96s)
            docker_cli_port_test.go:25: assertion failed: error is not nil: |:::9876!=[::]:9876|

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on integration-cli/docker_cli_port_test.go:
```diff
@@ -161,7 +175,7 @@ func assertPortList(c *testing.T, out string, expected []string) error {
 	// of the CLI used an incorrect output format for mappings on IPv6 addresses
 	// for example, "80/tcp -> :::80" instead of "80/tcp -> [::]:80".
 	oldFormat := func(mapping string) string {
-		old := strings.Replace(mapping, "-> [", "-> ", 1)
+		old := strings.Replace(mapping, "[", "", 1)
```
I made the normalisation too strict in https://github.com/moby/moby/pull/42050, and forgot the "expected" format was sometimes passed as a "full" format (`80/tcp -> [::]:80`), and sometimes only the mapping (`[::]:80`); pushed this fix as a separate commit in this PR, but I can move to a separate PR if we think that makes more sense

thaJeztah commented:
ok this is all green now; also prepared a draft backport for 20.10 https://github.com/moby/moby/pull/42205
